### PR TITLE
Gantt chart

### PR DESCRIPTION
**Pull request by afshinfard**

### CODE REVIEW

**Comment by afshinfard** [comment]
an example run to test it:
`Rscript -e 'rmarkdown::render("/projects/btl_scratch/aafshinfard/projects/physlr/port_code/sqcos/physlr/data/profile-physlr.Rmd", "html_document", "/projects/btl_scratch/aafshinfard/projects/physlr/port_code/sqcos/physlr/data/test.prof.html", knit_root_dir="/projects/btl_scratch/aafshinfard/projects/physlr/port_code/sqcos/physlr/data", params=list(path="/projects/btl_scratch/aafshinfard/projects/physlr/port_code/sqcos/physlr/data/"))'`

you can change the path to any path you like to test it with other files.

**Comment by jowong4** [comment]
Can you add a makefile command to generate the Gantt chart?

**Comment by afshinfard** [comment]
Sure @jowong4, I'll just do it.

**Comment by afshinfard** [comment]
Pufff I was struggling to make the `makefile` command work! and the problem was the formatting of spaces and newlines!

Now it works,
try ` make {NAME}.prof.html path_profile={PATH}` and use a name prefix you like and determine the path for `*.time` files. by defaults it is set to `data` path for physlr

**Comment by jowong4** [comment]
Hmm, are the ram values correct? It doesn't quite look correct.

**Comment by jowong4** [comment]
I would suggest making the current directory the default directory. You can probably do that by assining pwd to a variable and using that.

**Comment by afshinfard** [comment]
@jowong4 The ram values I extract from the `.time` file, so in case of python with multiprocessing it's the ram usage for only a single process, I can simply fix that by multiplying the ram usage with number of threads if the language is python, but I think it's better not to scale it up so we can see other bars easily.
also note that ram usage is converted into GB and also rounded (ceiling).
So is there still any other problem you see with the reports?

File: data/profile-physlr.Rmd [hunk 1]
```diff
@@ -0,0 +1,384 @@
+---
+title: "Physlr profiling - time and memory"
+author: "Amirhossein Afshinfard"
+date: "2/11/2020"
```
Comment (lcoombe) [inline data/profile-physlr.Rmd]:
```suggestion
```

File: data/profile-physlr.Rmd [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+cat(ldir)
+
+finf <- file.info(dir(path = ldir, full.names = TRUE), extra_cols = FALSE)
+sorted_finf <- finf[with(finf, order(as.POSIXct(ctime))), ]
+sorted_files <- rownames(sorted_finf)
+sorted_time_files <- sorted_files[substr(sorted_files,
+                                         nchar(sorted_files)-4,nchar(sorted_files)) == ".time"]
+
+# Or alternatively:
+#file.info("/projects/btl/jowong/github/physlr/data_hg004_stlfr/v2release/hg004_k40.histogram.time")
```
Comment (lcoombe) [inline data/profile-physlr.Rmd]: Was this an alternate approach you tried and decided against? We just don't want to have commented-out code

**Comment by jowong4** [comment]
Just change the green like we suggested in the meeting. Thanks!

File: data/profile-physlr.Rmd [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+  len = nchar(time_line)
+  colon_marks = str_match_all(time, regex(':',
+                                          ignore.case = TRUE, perl = TRUE))[[1]]
+
+  if (dim(colon_marks)[1] == 0)
+  {
+    cat(paste("Unexpected format"))
+  } else if (dim(colon_marks)[1] == 2)
+  {
+    df[i, "time_sec"] = strtoi(substr(time, nchar(time) - 1, nchar(time)), base =
```
Comment (lcoombe) [inline data/profile-physlr.Rmd]: Couldn't you both check which format the time string conforms to and extract the numbers in one go?

```
> n
[1] "10:40:22"
> str_match(n, regex('([:digit:]+)\\:([:digit:]+)\\:([:digit:]+)'))
     [,1]       [,2] [,3] [,4]
[1,] "10:40:22" "10" "40" "22"
```

File: data/profile-physlr.Rmd [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+
+    if (length(to_be_removed2) > 1) {
+      to_be_removed2 = to_be_removed2[2:length(to_be_removed2)]
+      cat("Following file(s) will be excluded due to having NA values")
+      print(print(sorted_time_files[to_be_removed2]))
+      print(df[to_be_removed2,])
+      df = df[!to_be_removed2,]
+    }
+
+    if (dim(df)[1] == 0) {
```
Comment (lcoombe) [inline data/profile-physlr.Rmd]: Could we just check this at the end? Seems like it is repeated a lot

Comment (afshinfard) [reply data/profile-physlr.Rmd]: I keep changing the `df` dataframe, so these if-statements are checking the file for different issues. I believe that some cases would not happen, but I just wanted to take into account any possibility...

Comment (lcoombe) [reply data/profile-physlr.Rmd]: Hmm..But if it that statement was true at any point in this code, wouldn't it still be true in the end? And you seem to have the same error message for each of them?

Comment (afshinfard) [reply data/profile-physlr.Rmd]: Umm, yes I can change the message to be more specific about the problem but that's not the main point... The main reason to check it multiple times is that if the statement is true the first time, then the operations I accomplish in the `else` part are not valid and running them results in error... I mean I am avoiding run-time errors this way...

File: data/profile-physlr.Rmd (hunk 3 above)
Comment (afshinfard) [reply data/profile-physlr.Rmd]: the two formats are different... How do I detect the format and extract the numbers in one go?
format 1: `HH:MM:SS`
format 2: `MM:SS.MS` which has only one `:` and also a `.`

Comment (lcoombe) [reply data/profile-physlr.Rmd]: You just need two regex expressions with the capture groups - that will tell you which format the string conforms to. After you do the `str_match`, you can check each for being non-NA with if-statements. Then, once you've entered the appropriate condition, you have easy access to the numbers you need.
Roughly what I was thinking:
```
hms <- str_match(...)
msms <- str_match(...)
if (is.na(hms[1,1)) {
...Parse as h,m,s..
} else if (is.na(msms[1,1]) {
...Parse as mmssms...
}
```

File: data/profile-physlr.Rmd (hunk 4 above)
Comment (lcoombe) [reply data/profile-physlr.Rmd]: My point wasn't about the error messages needing to be different (I don't think that is necessary), but about the repeated code which isn't great practice. You know best all the conditions you are checking for and how realistic the situations are (I find this part of the code a bit hard to follow), so I suggest looking at possibilities for refactoring to make it more concise.

**Comment by lcoombe** [comment]
@afshinfard - If you had different physlr runs in the same directory, would this script be able to distinguish the runs for the Gantt chart?

**Comment by afshinfard** [comment]
@lcoombe In that situation, the code orders all the `*.time` files and if there are multiple files for a single step it will keep the one which is older. so if you have two different runs in a single directory it will report the first run. One simple way to profile different runs is to `mv` the files you want to a subdirectory and run this on that subdirectory.

File: data/profile-physlr.Rmd (hunk 3 above)
Comment (afshinfard) [reply data/profile-physlr.Rmd]: I though you mean with a single call to `str_match()`. I just changed it to something like this.

File: data/profile-physlr.Rmd (hunk 4 above)
Comment (afshinfard) [reply data/profile-physlr.Rmd]: Sorry I was not good at answering this...
I just reformatted the code and now it checks once for all issues and then tries to remove rows with `NA` values.
